### PR TITLE
[3849] Trees in forms can display the same element many times

### LINKED ISSUE #3849: Related elements view can crash with MUI 5
* [x] **I have checked that this bug has not yet been reported by someone else**
* [x] **I have checked that this bug appears on Chrome**
* [x] **I have specified the version** :
- 2024.7.9
* [x] **I have specified my environment** : 
- All

### Stack

```
Error: MUI X: The Tree View component requires all items to have a unique `id` property.
Alternatively, you can use the `getItemId` prop to specify a custom id for each item.
Two items were provided with the same id in the `items` prop: "cf164a01-97ad-4ab1-9ef3-8874dfadbbbe"
    at index-DmZyy1EZ.js:2578:930
    at bS (index-DmZyy1EZ.js:52:54278)
    at eK (index-DmZyy1EZ.js:52:54733)
    at Object.useState (index-DmZyy1EZ.js:52:60651)
    at pl.useState (index-DmZyy1EZ.js:12:6151)
    at n$r (index-DmZyy1EZ.js:2574:7579)
    at index-DmZyy1EZ.js:2580:1637
    at YUe (index-DmZyy1EZ.js:52:53496)
    at yBt (index-DmZyy1EZ.js:52:61687)
    at cnn (index-DmZyy1EZ.js:56:10829)
jBe @ index-DmZyy1EZ.js:54
```

### Steps to reproduce

1/ Use the `Sirius Web` template project
2/ Activate the "Related elements" view
3/ Select the following object:
Capture d’écran 2024-08-08 à 15 39 26

The crash occurs

### Expected behavior

No crash

### Actual behavior

Crash

### PR DESCRIPTION
**Pull request by gcoutable**
# Pull request template

## General purpose
What is the main goal of this pull request?
- [x] Bug fixes
- [ ] New features
- [ ] Documentation
- [ ] Cleanup
- [ ] Tests
- [ ] Build / releng

## Project management
- [x] Has the pull request been added to the relevant project and milestone? (Only if you know that your work is part of a specific iteration such as the current one)
- [x] Have the `priority:` and `pr:` labels been added to the pull request? (In case of doubt, start with the labels `priority: low` and `pr: to review later`)
- [x] Have the relevant issues been added to the pull request?
- [x] Have the relevant labels been added to the issues? (`area:`, `difficulty:`, `type:`)
- [x] Have the relevant issues been added to the same project and milestone as the pull request?
- [x] Has the `CHANGELOG.adoc` been updated to reference the relevant issues?
- [ ] Have the relevant API breaks been described in the `CHANGELOG.adoc`? (Including changes in the GraphQL API)
- [ ] In case of a change with a visual impact, are there any screenshots in the `CHANGELOG.adoc`? For example in `doc/screenshots/2022.5.0-my-new-feature.png`

## Architectural decision records (ADR)
- [ ] Does the title of the commit contributing the ADR start with `[doc]`?
- [ ] Are the ADRs mentioned in the relevant section of the  `CHANGELOG.adoc`?

## Dependencies
- [ ] Are the new / upgraded dependencies mentioned in the relevant section of the `CHANGELOG.adoc`?
- [ ] Are the new dependencies justified in the `CHANGELOG.adoc`?

## Frontend

This section is not relevant if your contribution does not come with changes to the frontend.

### General purpose
- [ ] Is the code properly tested? (Plain old JavaScript tests for business code and tests based on React Testing Library for the components)

### Typing
We need to improve the typing of our code, as such, we require every contribution to come with proper TypeScript typing for both changes contributing new files and those modifying existing files.
Please ensure that the following statements are true for each file created or modified (this may require you to improve code outside of your contribution).

- [ ] Variables have a proper type
- [ ] Functions’ arguments have a proper type
- [ ] Functions’ return type are specified
- Hooks are properly typed:
	- [ ] `useMutation<DATA_TYPE, VARIABLE_TYPE>(…)`
	- [ ] `useQuery<DATA_TYPE, VARIABLE_TYPE>(…)`
	- [ ] `useSubscription<DATA_TYPE, VARIABLE_TYPE>(…)`
	- [ ] `useMachine<CONTEXT_TYPE, EVENTS_TYPE>(…)`
	- [ ] `useState<STATE_TYPE>(…)`
- [ ] All components have a proper typing for their props
- [ ] No useless optional chaining with `?.` (if the GraphQL API specifies that a field cannot be `null`, do not treat it has potentially `null` for example)
- [ ] Nullable values have a proper type (for example `let diagram: Diagram | null = null;`)

## Backend

This section is not relevant if your contribution does not come with changes to the backend.

### General purpose
- [ ] Are all the event handlers tested?
- [ ] Are the event processor tested?
- [ ] Is the business code (services) tested?
- [ ] Are diagram layout changes tested?

### Architecture
- [ ] Are data structure classes properly separated from behavioral classes?
- [ ] Are all the relevant fields final?
- [ ] Is any data structure mutable? If so, please write a comment indicating why
- [ ] Are behavioral classes either stateless or side effect free?

### How to test this PR?
_Please describe here the various use cases to test this pull request_

- [ ] Has the Kiwi TCMS test suite been updated with tests for this contribution?

### CODE REVIEW

**Comment by pcdavid** [comment]
I can still trigger the crash with this. It's difficult to know on which element exactly because I have to try several and once it crashes it's too late.
Probably a case where the same element appears multiple times *even inside the same parent*.
Maybe the child's index/position should also be used to compute the id?

**Comment by pcdavid** [comment]
Actually I can reproduce it on the "Related Elements" of this item of the "Sirius Web" Papaya example:

image

File: packages/forms/frontend/sirius-components-forms/src/propertysections/TreePropertySection.tsx [hunk 1]
```diff
@@ -142,14 +149,18 @@ const TreeItem = ({ node, nodes, readOnly, editingContextId, formId, widgetId }:
 
   const handleClick: React.MouseEventHandler<HTMLDivElement> = () => {
     if (node.selectable) {
+      const treePath = node.id.split('/');
+      const nodeId = treePath[treePath.length - 1];
       const newSelection: SelectionEntry = {
-        id: node.id,
+        id: nodeId,
```
Comment (sbegaudeau) [inline packages/forms/frontend/sirius-components-forms/src/propertysections/TreePropertySection.tsx]: Why does this need to change? I don't think that you are changing the property of the node object in this PR unless I'm missing something?

**Comment by sbegaudeau** [comment]
> Actually I can reproduce it on the "Related Elements" of this item of the "Sirius Web" Papaya example:
> 
> image

Could it be that one element could appear with the same path twice? Given the nature of Papaya which can have duplicates in some references it's not impossible. We may have to add the index of the object in the list of tree item to create a really unique id....

**Comment by pcdavid** [comment]
> Could it be that one element could appear with the same path twice? Given the nature of Papaya which can have duplicates in some references it's not impossible. We may have to add the index of the object in the list of tree item to create a really unique id....

Yes, that's what I said [here](https://github.com/eclipse-sirius/sirius-web/pull/3883#issuecomment-2311721702) :)

**Comment by sbegaudeau** [comment]
> > Could it be that one element could appear with the same path twice? Given the nature of Papaya which can have duplicates in some references it's not impossible. We may have to add the index of the object in the list of tree item to create a really unique id....
> 
> Yes, that's what I said [here](https://github.com/eclipse-sirius/sirius-web/pull/3883#issuecomment-2311721702) :)

I missed that sorry -.- 

😄

**Review by sbegaudeau — APPROVED** [review]
I'll test it locally while waiting for the build but it looks good

**Pull request merged**